cmd/aws-s3-csi-controller/csicontroller/reconciler.go

```diff
@@ -163,56 +177,269 @@ func (r *Reconciler) spawnOrDeleteMountpointPodIfNeeded(
+					"mountpointPodName", mpPodName)
+				return true, nil
+			}
 		}
+	}
 
-		// No need to do anything - either there was no Mountpoint Pod for `pod` or it was in `Running` state,
-		// so a clean unmount operation will be performed and Mountpoint Pod will cleany exit (and get deleted by `reconcileMountpointPod`).
-		return nil
+	// Delete MountpointS3PodAttachment if map is empty
+	if len(s3pa.Spec.MountpointS3PodToWorkloadPodUIDs) == 0 {
+		log.Info("MountpointS3PodAttachment has zero Mountpoint Pods. Will delete it")
+		err := r.Delete(ctx, s3pa)
+		if apierrors.IsConflict(err) {
+			log.Info("Failed to delete MountpointS3PodAttachment due to resource conflict, requeueing")
+			return true, nil
+		}
 	}
 
-	if isMountpointPodExists {
-		log.V(debugLevel).Info("Mountpoint Pod already exists - ignoring")
-		return nil
+	return false, nil
+}
+
+func (r *Reconciler) handleNewS3PodAttachment(
+	ctx context.Context,
+	workloadPod *corev1.Pod,
+	pv *corev1.PersistentVolume,
+	fieldFilters client.MatchingFields,
+	log logr.Logger,
+) (bool, error) {
+	if r.s3paExpectations.IsPending(fieldFilters) {
+		log.Info("MountpointS3PodAttachment creation is pending, requeueing")
+		return true, nil
 	}
 
-	if err := r.spawnMountpointPod(ctx, workloadPod, pvc, pv, csiSpec, mpPodName); err != nil {
+	if err := r.createS3PodAttachmentWithMPPod(ctx, workloadPod, pv, log); err != nil {
+		return false, err
+	}
+
+	r.s3paExpectations.SetPending(fieldFilters)
+	return true, nil
+}
+
+func (r *Reconciler) createS3PodAttachmentWithMPPod(
+	ctx context.Context,
+	workloadPod *corev1.Pod,
+	pv *corev1.PersistentVolume,
+	log logr.Logger,
+) error {
+	authSource := r.getAuthSource(pv)
+	mpPodName, err := r.spawnMountpointPod(ctx, workloadPod, pv, log)
+	if err != nil {
 		log.Error(err, "Failed to spawn Mountpoint Pod")
 		return err
 	}
 
+	fsGroup := ""
+	if workloadPod.Spec.SecurityContext.FSGroup != nil {
+		fsGroup = strconv.FormatInt(*workloadPod.Spec.SecurityContext.FSGroup, 10)
+	}
+	s3pa := &crdv1.MountpointS3PodAttachment{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "s3pa-",
+			Labels: map[string]string{
+				LabelCSIDriverVersion: r.mountpointPodConfig.CSIDriverVersion,
+			},
+		},
+		Spec: crdv1.MountpointS3PodAttachmentSpec{
+			NodeName:             workloadPod.Spec.NodeName,
+			PersistentVolumeName: pv.Name,
+			VolumeID:             pv.Spec.CSI.VolumeHandle,
+			MountOptions:         strings.Join(pv.Spec.MountOptions, ","),
+			WorkloadFSGroup:      fsGroup,
+			AuthenticationSource: authSource,
+			MountpointS3PodToWorkloadPodUIDs: map[string][]string{
+				mpPodName: {string(workloadPod.UID)},
+			},
+		},
+	}
+	if authSource == "pod" {
+		s3pa.Spec.WorkloadNamespace = workloadPod.Namespace
+		s3pa.Spec.WorkloadServiceAccountName = getServiceAccountName(workloadPod)
+
+		roleARN, err := r.findIRSAServiceAccountRole(ctx, workloadPod)
+		if err != nil {
+			return err
```

Comment:
Maybe we should perform this check before spawning MP Pod? If we run into this error, it would leave things in partial state (i.e., there is an MP Pod but no attachment)